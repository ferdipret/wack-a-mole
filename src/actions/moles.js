import {
  UPDATE_ACTIVE_MOLES,
  UPDATE_WACKED_MOLE,
  CLEAR_ACTIVE_MOLES,
} from "../constants"

export function placeMolesInField(payload) {
  return {
    type: UPDATE_ACTIVE_MOLES,
    payload,
  }
}

export function updateWackedMole(payload) {
  return {
    type: UPDATE_WACKED_MOLE,
    payload,
  }
}

export function clearMolesInField() {
  return {
    type: CLEAR_ACTIVE_MOLES,
  }
}
