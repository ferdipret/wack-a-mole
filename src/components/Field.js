import React from "react"
import PropTypes from "prop-types"
import styled from "styled-components"
import Mole from "./Mole"
import Overlay from "./Overlay"

const FieldWrapper = styled.div`
  background: hsl(120, 50%, 70%);
  display: flex
  font-family: 'Share Tech Mono', monospace;
  flex-wrap: wrap;
  position: relative;
`

function Field({moles, onMoleClick, running, timer, score}) {
  const getOverlayState = t =>  {
    switch(t) {
      case 60: {
        return <span>Click the `START` button to begin a game</span>
      }

      case 0: {
        return (
          <div>
            <h2>GAME OVER!</h2>
            <h3>score</h3>
            <h1>{score}</h1>
          </div>
        )
      }

      default: {
        return <span>Game Paused</span>
      }
    }
  }
  return (
    <FieldWrapper>
      {!running && <Overlay>{getOverlayState(timer)}</Overlay>}
      {moles.map(mole => (
        <Mole key={mole.index} {...mole} onMoleClick={onMoleClick}/>
      ))}
    </FieldWrapper>
  )
}

Field.propTypes = {
  moles: PropTypes.arrayOf(PropTypes.object),
  onMoleClick: PropTypes.func.isRequired,
  running: PropTypes.bool.isRequired,
  timer: PropTypes.number.isRequired,
  score: PropTypes.number.isRequired,
}

export default Field
