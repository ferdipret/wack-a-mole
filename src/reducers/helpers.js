export function getRandomHoles() {
  const getValue = () => Math.floor(Math.random() * 16)

  let array = []
  do {
    const newValue = getValue()
    if(array.indexOf(newValue) === -1) {
      array = array.concat(newValue)
    }
  } while (array.length !== 4) 

  return array
}

export function calculateScore(hits)  {
  // Not so much a calculation but rather a get
  switch(hits) {
    case 1: return 1
    case 2: return 4
    case 3: return 16
    case 4: return 256
    default: return 0
  }
}
