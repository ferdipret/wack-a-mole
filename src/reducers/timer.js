import { UPDATE_TIMER } from "../constants"

const timer = (state = 60, action) => {
  let newState

  switch (action.type) {
    case UPDATE_TIMER: {
      newState = action.payload
      break
    }

    default: {
      newState = state
      break
    }
  }

  return newState
}

export default timer
