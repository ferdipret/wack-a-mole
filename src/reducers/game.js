import {
  UPDATE_RUNNING_STATE,
  UPDATE_GAME_PHASE,
  GAME_OVER,
  SET_WACK_COUNT,
  SET_SCORE,
  RESET_GAME,
} from "../constants"

import { calculateScore } from "./helpers"

export const initialState = {
  running: false,
  phase: 1,
  gameover: false,
  uptimeWackCount: 0,
  score: 0,
  bombCount: 4,
}

const game = (state = initialState, action) => {
  let newState

  switch (action.type) {
    case GAME_OVER: {
      newState = {
        ...state,
        gameover: action.payload,
      }
      break
    }

    case RESET_GAME: {
      newState = initialState
      break
    }

    case UPDATE_RUNNING_STATE: {
      newState = {
        ...state,
        running: action.payload,
      }
      break
    }

    case SET_WACK_COUNT: {
      newState = {
        ...state,
        uptimeWackCount: action.payload,
      }
      break
    }

    case SET_SCORE: {
      newState = {
        ...state,
        uptimeWackCount: 0,
        score: state.score + calculateScore(action.payload.uptimeWackCount),
        bombCount: action.payload.addBomb
          ? state.bombCount - 1
          : state.bombCount,
      }
      break
    }

    case UPDATE_GAME_PHASE: {
      newState = {
        ...state,
        phase: state.phase + 1,
      }
      break
    }

    default: {
      newState = state
      break
    }
  }

  return newState
}

export default game
