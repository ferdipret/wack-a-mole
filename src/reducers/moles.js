import {
  UPDATE_ACTIVE_MOLES,
  UPDATE_WACKED_MOLE,
  CLEAR_ACTIVE_MOLES,
} from "../constants"
import { getRandomHoles } from "./helpers"

export const initialState = new Array(16).fill(undefined).map((mole, index) => {
  return {
    index,
    active: false,
    bomb: false,
    hidingTime: 55,
  }
})

const moles = (state = initialState, action) => {
  let newState

  switch (action.type) {
    case UPDATE_ACTIVE_MOLES: {
      const newHolesArray = getRandomHoles()
      newState = action.payload.moles.map(mole => ({
        ...mole,
        active:action.payload.addBomb &&
          newHolesArray.indexOf(mole.index) === 3
          ?  false
          : newHolesArray.indexOf(mole.index) !== -1,
        hidingTime: action.payload.uptime,
        bomb: action.payload.addBomb && newHolesArray[3] === mole.index,
      }))
      break
    }

    case CLEAR_ACTIVE_MOLES: {
      newState = state.map(mole => ({
        ...mole,
        active: false,
      }))
      break
    }

    case UPDATE_WACKED_MOLE: {
      newState = state.map(mole => ({
        ...mole,
        active: mole.index === action.payload ? false : mole.active,
      }))
      break
    }

    default: {
      newState = state
      break
    }
  }

  return newState
}

export default moles
