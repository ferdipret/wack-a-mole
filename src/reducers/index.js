import { combineReducers } from "redux"
import timer from "./timer"
import game from "./game"
import moles from "./moles"

export default combineReducers({ timer, game, moles })
