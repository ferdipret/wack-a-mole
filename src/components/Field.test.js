import React from "react"
import { shallow } from "enzyme"
import Field from "./Field"
import { initialState } from "../reducers/moles"

describe("Field", () => {
  const onMoleClick = jest.fn()
  const props = {
    moles: initialState,
    onMoleClick,
    running: false,
    timer: 0,
    score: 0,
  }
  const wrapper = shallow(<Field {...props} />)

  it("renders properly", () => {
    expect(wrapper).toMatchSnapshot()
  })
})
