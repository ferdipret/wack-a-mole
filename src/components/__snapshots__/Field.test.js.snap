// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field renders properly 1`] = `
<styled.div>
  <styled.div>
    <div>
      <h2>
        GAME OVER!
      </h2>
      <h3>
        score
      </h3>
      <h1>
        0
      </h1>
    </div>
  </styled.div>
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={0}
    key="0"
    onMoleClick={[Function]}
  />
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={1}
    key="1"
    onMoleClick={[Function]}
  />
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={2}
    key="2"
    onMoleClick={[Function]}
  />
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={3}
    key="3"
    onMoleClick={[Function]}
  />
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={4}
    key="4"
    onMoleClick={[Function]}
  />
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={5}
    key="5"
    onMoleClick={[Function]}
  />
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={6}
    key="6"
    onMoleClick={[Function]}
  />
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={7}
    key="7"
    onMoleClick={[Function]}
  />
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={8}
    key="8"
    onMoleClick={[Function]}
  />
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={9}
    key="9"
    onMoleClick={[Function]}
  />
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={10}
    key="10"
    onMoleClick={[Function]}
  />
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={11}
    key="11"
    onMoleClick={[Function]}
  />
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={12}
    key="12"
    onMoleClick={[Function]}
  />
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={13}
    key="13"
    onMoleClick={[Function]}
  />
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={14}
    key="14"
    onMoleClick={[Function]}
  />
  <Mole
    active={false}
    bomb={false}
    hidingTime={55}
    index={15}
    key="15"
    onMoleClick={[Function]}
  />
</styled.div>
`;
